GDT.md

```diff
@@ -0,0 +1,220 @@
+
+Accessing memory inside a segment is done relative to it's base. Lets say you have a segment with a base of 0x1000,
+and some data in memory at address 0x1100.
+The data would be accessed at address 0x100 (assuming the segment is the active DS), as addressed are translated as `segment_base + offset`. In this case the segment base is 0x1000, and the offset is 0x100.
+
+Segments can also be explicitly referenced. To load something at offset 0x100 into the ES region, you can use `mov es:0x100, $rax`. This would perform the translation from logical address to linear address using ES instead of DS (the default for data), a common example is when an interrupt occurs while the cpu is in ring 3, it will switch to ring 0 and load the appropriate descriptors into the segment registers.
+
+### Segment Registers
+
+The various segment registers and their uses are outline below. There are some tricks to load a descriptor from the GDT into a segment register. They can't be mov'd into directly, so you'll need to use a scratch register to change their value. The cpu will also automatically reload segment registers on certain events (see the manual for these). 
+
+To load any of the data registers, use the following:
+```
+#at&t syntax
+
+#example: load ds with the first descriptor
+mov $0x8, %ax       #any register will do, ax is used for the example here
+mov %ax, %ds
+
+#example: load ss with second descriptor
+mov $0x10, %ax
+mov %ax, %ss
+```
+
+Changing CS (code segment) is a little trickier, as it cant be written to directly, instead it requires a far jump:
+```
+#at&t syntax
+
+reload_cs:
+    pop %rdi        #call instruction pushes return address on to stack, preserve it in rdi (or edi if 32 bit)
+    push $0x8       #the code selector is 0x8 in this case (the first non-null descriptor in the GDT)
+    push %rdi       #the stack now looks like cs:rip, which is what a far jump expects.
+    retfq           #pops the cs:rip from the stack, loads the new code segment, and begins executing at that address.
+```
+
+## Segmentation and Paging
+
+When segmentation and paging are used together, segmentation is applied first, then paging.
+The process of translation an address is as follows:
+- Calculate linear address: `logical_address + segment_base`.
+- Traverse paging structure for physical address, using linear address.
+- Access memory at physical address.
+
+## Using the GDT
+
+All the theory is great, but how to apply it? 
+A simple example is outline just below, for a simple 64-bit long mode setup you'd need
+- Selector 0x00: null
+- Selector 0x08: kernel code (64-bit, ring 0)
+- Selector 0x10: kernel data (64-bit)
+- Selector 0x18: user code (64-bit, ring 3)
+- Selector 0x20: user data (64-bit)
+
+To populate these entries, we'll use the following:
```

Comment:
I would say before all the code why not place a table that explains all of the fields in a gdt entry? 

And if needed also another table that explains the bits of the address loaded with lgdt

Reply:
Added a table and more details on the contents of the commonly used GDT descriptors.